drivers/bluetooth/hci/hci_sf32lb.c

```diff
@@ -0,0 +1,616 @@
+
+	while (1) {
+		int len;
+
+		k_sem_take(&(data->sem), K_FOREVER);
+		len = ipc_queue_get_rx_size(data->ipc_port);
+		while (len) {
+			struct h4_data *h4 = dev->data;
+			struct net_buf *buf;
+
+			LOG_DBG("%s len %d", __func__, len);
+			process_rx(dev);
+			buf = k_fifo_get(&h4->rx.fifo, K_NO_WAIT);
+			while (buf) {
+				LOG_DBG("Calling bt_recv(%p),len=%d,data=%p", (void *)buf, buf->len,
+					(void *)buf->data);
+
+				if (buf->len == 0 || buf->data == NULL) {
+					break;
+				}
+
+				if (h4->recv) {
+					sf32lb_hci_trace(buf->data[0], &buf->data[1], buf->len - 1,
+							 H4TL_PACKET_CTRL);
+					h4->recv(dev, buf);
+				} else {
+					net_buf_unref(buf);
+				}
+
+				buf = k_fifo_get(&h4->rx.fifo, K_NO_WAIT);
+			};
+			len = ipc_queue_get_rx_size(data->ipc_port);
+		}
+		process_tx(dev);
+	}
+}
+
+static size_t h4_discard(const struct device *dev, size_t len)
+{
+	struct h4_data *h4 = dev->data;
+	uint8_t buf[33];
+	int err;
+
+	err = ipc_queue_read(h4->ipc_port, buf, MIN(len, sizeof(buf)));
+	if (unlikely(err < 0)) {
+		LOG_ERR("Unable to read from UART (err %d)", err);
+		err = 0;
+	}
+	return err;
+}
+
+static inline void read_payload(const struct device *dev)
+{
+	struct h4_data *h4 = dev->data;
+	struct net_buf *buf;
+	int read;
+
+	if (!h4->rx.buf) {
+		size_t buf_tailroom;
+
+		h4->rx.buf = get_rx(h4, K_NO_WAIT);
+		if (!h4->rx.buf) {
+			if (h4->rx.discardable) {
+				LOG_WRN("Discarding event 0x%02x", h4->rx.evt.evt);
+				h4->rx.discard = h4->rx.remaining;
+				reset_rx(h4);
+				return;
+			}
+			LOG_WRN("Failed to allocate, deferring to rx_thread");
+			return;
+		}
+		LOG_DBG("Allocated rx.buf %p", (void *)h4->rx.buf);
+
+		buf_tailroom = net_buf_tailroom(h4->rx.buf);
+		if (buf_tailroom < h4->rx.remaining) {
+			LOG_ERR("Not enough space in buffer %u/%zu", h4->rx.remaining,
+				buf_tailroom);
+			h4->rx.discard = h4->rx.remaining;
+			reset_rx(h4);
+			return;
+		}
+		copy_hdr(h4);
+	}
+
+	read = ipc_queue_read(h4->ipc_port, net_buf_tail(h4->rx.buf), h4->rx.remaining);
+	if (unlikely(read < 0)) {
+		LOG_ERR("Failed to read UART (err %d)", read);
+		return;
+	}
+
+	net_buf_add(h4->rx.buf, read);
+	h4->rx.remaining -= read;
+
+	LOG_DBG("got %d bytes, remaining %u", read, h4->rx.remaining);
+	LOG_DBG("Payload (len %u): %s", h4->rx.buf->len, bt_hex(h4->rx.buf->data, h4->rx.buf->len));
+
+	if (h4->rx.remaining) {
+		return;
+	}
+
+	buf = h4->rx.buf;
+	h4->rx.buf = NULL;
+
+	reset_rx(h4);
+
+	LOG_DBG("Putting buf %p to rx fifo", (void *)buf);
+	k_fifo_put(&h4->rx.fifo, buf);
+}
+
+static inline void read_header(const struct device *dev)
+{
+	struct h4_data *h4 = dev->data;
+
+	switch (h4->rx.type) {
+	case BT_HCI_H4_NONE:
+		h4_get_type(dev);
+		return;
+	case BT_HCI_H4_EVT:
+		get_evt_hdr(dev);
+		break;
+	case BT_HCI_H4_ACL:
+		get_acl_hdr(dev);
+		break;
+	case BT_HCI_H4_SCO:
+		if (IS_ENABLED(CONFIG_BT_CLASSIC)) {
+			get_sco_hdr(dev);
+		} else {
+			LOG_ERR("SCO got unexpected\n");
+			CODE_UNREACHABLE;
+		}
```

Comment:
`CODE_UNREACHABLE;` doesn't seem correct, since it's theoretically possible to reach it if there's some corrupt data or similar. If you follow my earlier proposal of letting this go to `default:` then that takes care of it.